[Serializer] Name converter support

dunglas opened the pull request:
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | yes |
| Tests pass? | yes |
| Fixed tickets | #12212 |
| License | MIT |
| Doc PR | symfony/symfony-docs#4692 |

This PR adds support for custom property naming strategies to the serializer and provides a built-in NameConverter using this new system: (CamelCase to underscore).
It handles normalization and denormalization (convert `fooBar` to `foo_bar` when serializing, then from `foo_bar` to `fooBar` when deserializing). It also has a flag to convert only some attributes.

The `setCamelizedAttributes()` is deprecated in favor of this new method (more flexible, allows to rename all attributes of a class and support deserialization) and now uses it internally.

dunglas committed:
[Serializer] Name converter support

dunglas commented:
Should be merged only after #13252 merge (and will need a rebase).

fabpot commented:
@dunglas #13252 is merged now, can you rebase?

fabpot commented on src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php:
```diff
@@ -176,23 +192,6 @@ protected function handleCircularReference($object)
-     * @param string $attributeName
-     * @return string
-     */
-    protected function formatAttribute($attributeName)
```
That's a BC break.

dunglas commented:
@fabpot BC break reverted.

xabbuh commented on src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php:
```diff
@@ -178,18 +194,17 @@ protected function handleCircularReference($object)
-                return ('.' === $match[1] ? '_' : '').strtoupper($match[2]);
-            }, $attributeName);
-        }
+        trigger_error('The formatAttribute() method has been deprecated. Use Symfony\Component\Serializer\NameConverter\CamelCaseToSnakeCaseNameConverter instead.', E_USER_DEPRECATED);
```
Using `__METHOD__` would be better than using `formatAttribute()` (otherwise you don't know the method's class).

xabbuh replied:
and the wording actually is a bit different: http://symfony.com/doc/current/contributing/code/conventions.html#deprecations

dunglas commented:
@xabbuh fixed

xabbuh commented:
@dunglas You should use the actual version numbers instead of `2.X` and `2.Y`. ;)

dunglas commented:
Oops :)

dunglas commented:
fixed

fabpot commented:
:+1:

stof commented on src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php:
```diff
@@ -114,13 +119,24 @@ public function setIgnoredAttributes(array $ignoredAttributes)
+            }, $camelizedAttribute));
+        }
+
+        $this->nameConverter = new CamelCaseToSnakeCaseNameConverter($attributes);
```
this will break if a NameConverter was passed in the constructor (you will removed the existing one entirely)

dunglas replied:
This is intended.
As NameConverter deprecate this method, both must not be used together. I'll add the throwing of an exception in that case (I thought it was done, but this is not the case).

dunglas replied:
Exception added.

stof commented:
Please rebase this to fix conflicts (the logic common to the GetSetNormalizer and the PropertyNormalizer has been moved to their common parent class)

dunglas commented:
Rebased. Thanks @stof for the review.

dunglas commented:
Tests fixed.

fabpot commented:
@dunglas Looks like the tests are not fixed (at least, according to Travis)

dunglas commented:
Strange, the last commit should fix that (it works locally). Can you run Travis again?

dunglas commented:
@fabpot problem with Travis IMO. Commits squashed btw.

fabpot commented:
@dunglas Tests pass now on Travis as well apparently.

stof commented:
The changelog of the component should be updated to mention the changes being done (adding the new feature and deprecating the old one). Otherwise :+1:

dunglas committed:
[Serializer] Update changelog

dunglas commented:
Updated the changelog with new features of 2.6. I've also added a new feature of 2.6 missing from the changelog. Should I create a separate PR on the 2.6 branch for that?

fabpot commented on src/Symfony/Component/Serializer/CHANGELOG.md:
```diff
@@ -1,11 +1,26 @@
 
  * added a new serializer: `PropertyNormalizer`. Like `GetSetMethodNormalizer`,
    this normalizer will map an object's properties to an array.
+ * added circular references handling for `GetSetMethodNormalizer`
```
Indeed, you should submit a PR on 2.6 for this one.

fabpot commented:
Thank you @dunglas.
